winml/lib/Api/impl/Tensor.h

```diff
@@ -15,79 +15,78 @@ namespace _winml {
 template <typename T>
 class Tensor {
  private:
-  using TensorBuffer = TensorBuffer<T>;
-  using TensorBufferPtr = typename TensorBuffer::TensorBufferPtr;
-
-  TensorBufferPtr m_buffer;
+  std::shared_ptr<TensorBuffer<T>> buffer_;
   std::vector<int64_t> shape_;
 
  public:
   Tensor() = delete;
 
   Tensor(
       std::vector<int64_t> const& shape,
-      wss::IBuffer buffer) : shape_(shape),
-                             m_buffer(
-                                 TensorBuffer::Create(
-                                     static_cast<uint32_t>(
-                                         std::accumulate(
-                                             std::begin(shape),
-                                             std::end(shape),
-                                             static_cast<int64_t>(1),
-                                             std::multiplies<int64_t>())),
-                                     buffer)) {
-  }
+      wfc::IIterable<wss::IBuffer> const& buffers) :
+                            shape_(shape),
+                            buffer_(TensorBuffer<T>::Create(
+                                        static_cast<size_t>(std::accumulate(
+                                             std::begin(shape), std::end(shape),
+                                             static_cast<int64_t>(1), std::multiplies<int64_t>())),
+                                        buffers)) {}
 
   Tensor(
       std::vector<int64_t> const& shape) : shape_(shape),
-                                           m_buffer(
-                                               TensorBuffer::Create(
-                                                   static_cast<uint32_t>(
-                                                       std::accumulate(
-                                                           std::begin(shape),
-                                                           std::end(shape),
-                                                           static_cast<int64_t>(1),
-                                                           std::multiplies<int64_t>())))) {
-  }
+                                           buffer_(TensorBuffer<T>::Create(
+                                                        static_cast<size_t>(std::accumulate(
+                                                            std::begin(shape), std::end(shape),
+                                                            static_cast<int64_t>(1),
+                                                            std::multiplies<int64_t>())))) {}
```

Comment:
Is there a reason to do the math in int64_t and cast to size_t at the end? I feel like it still doesn't work if there's overflow in x86

Reply:
It is only int64_t because int64_t is the onnx standard for the dimension (-1) allowed for shape descriptions (but not allows for actual tensors)... Our apis expect int64_t for that reason, and thats why the shape is int64_t. The size_t cast is done to ensure that we dont have actual negative size buffers. For shapes, this is not a problem practical, as they should not be that big.

Reply:
Then isn't

```c++
std::accumulate(std::begin(shape), std::end(shape), static_cast<size_t>(1), std::multiplies<size_t>()))
```

Equivalent to

```c++
static_cast<size_t>(std::accumulate(std::begin(shape), std::end(shape), static_cast<int64_t>(1), std::multiplies<int64_t>()))
```

?

Reply:
I believe you cannot accumulate an iterable of int64_t with a start, and multiplication of size_t. I think that was hit when the implementation was first done, and thus the cast was added.

This change isnt changing that behavior.